concent_api/api-integration-multi-threads-test.py

```diff
@@ -0,0 +1,214 @@
+
+# !/usr/bin/env python3
+import os
+import sys
+import time
+from threading import Thread
+from typing import Optional
+
+import requests
+from golem_messages import message
+from golem_messages.factories.tasks import ReportComputedTaskFactory
+from golem_messages.message import Message
+from golem_messages.message.concents import ForceGetTaskResult
+from golem_messages.message.concents import ForceReportComputedTask
+from golem_messages.message.tasks import AckReportComputedTask
+from golem_messages.message.tasks import ReportComputedTask
+
+from api_testing_common import count_fails
+from api_testing_common import PROVIDER_PRIVATE_KEY
+from api_testing_common import REQUESTOR_PRIVATE_KEY
+from api_testing_common import api_request
+from api_testing_common import assert_condition
+from api_testing_common import create_signed_task_to_compute
+from api_testing_common import run_tests
+from api_testing_common import timestamp_to_isoformat
+from common.helpers import get_current_utc_timestamp
+from common.helpers import sign_message
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "concent_api.settings")
+
+REPORT_COMPUTED_TASK_SIZE = 10
+
+NUMBER_OF_TESTING_THREADS = 3  # changing this variable remember to correct number of expected responses
```

Comment:
Please add assert to this statement

Reply:
It is not good idea. At the end of test I compare list of responses from views with `expected list`- `expected list` is hardcoded in test. E.g. if you are changing number of threads from 3 to 2, you should change `expected_responses = ['AckForceGetTaskResult', 'ServiceRefused', 'ServiceRefused']` to `expected_responses = ['AckForceGetTaskResult', 'ServiceRefused']`. 

Second thing is, if something gone wrong (test failed), number of responses may be incorrect. But it is not simple assert for numbers, it has to be validated. I added in error_message in this validation comment, that maybe test failed and maybe someone simple forgot to change number of threads    

This comment is to reme

Reply:
@pawelkisielewicz 

> It is not good idea. At the end of test I compare list of responses from views with `expected list`- `expected list` is hardcoded in test. E.g. if you are changing number of threads from 3 to 2, you should change `expected_responses = ['AckForceGetTaskResult', 'ServiceRefused', 'ServiceRefused']` to `expected_responses = ['AckForceGetTaskResult', 'ServiceRefused']`. 

But that's exactly why you need to use `assert`. It does not need to be here. You can put it in the test. If you add
``` python
assert len(expected_responses) == NUMBER_OF_TESTING_THREADS
```
you won't have to worry about someone changing the constant and not updating the list - python will be able to detect it and also anyone reading the code will see that the list length should match the number of threads. It serves both as documentation and as a method to enforce it.

> Second thing is, if something gone wrong (test failed), number of responses may be incorrect. But it is not simple assert for numbers, it has to be validated. I added in error_message in this validation comment, that maybe test failed and maybe someone simple forgot to change number of threads

Looks like you have just discovered why we need `assert` in tests and why it's different from test assertions ;)

I have already explained this in the past. Please see this comment: https://github.com/golemfactory/concent/pull/673#discussion_r205289860

Simply put:
- Use `assert` to check if the test itself is correct and it it makes sense to even launch it (i.e. if the list contains expected value for every thread).
- Use test assertions (`assertEqual`, etc.) to check if the results returned by the code you're testing are correct (i.e. if the threads returned correct values).

Reply:
Thanks, done :)